### PR TITLE
[FIX] web: useNumpadDecimal should take into account the selection

### PR DESCRIPTION
**Pull request by caburj**
**[FIX] web: useNumpadDecimal should take into account the selection**

When NumpadDecimal is triggered, appending the decimalPoint to the
input field value is not enough, rather, we should replace the selected
range.

Also, we should ignore this for input field of type="number" because
it's natively taken into account.

**[FIX] web: useNumpadDecimal: synchronize the inputs listeners**

This makes sure that the handler is properly assigned/removed
to/from the added/removed input elements.

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/web/static/tests/views/fields/numeric_fields_tests.js [hunk 1]
```diff
@@ -161,6 +161,8 @@ QUnit.module("Fields", (hooks) => {
 
             await click(target.querySelector(".o_progress"));
             const progressbarInputs = target.querySelectorAll(".o_field_progressbar input");
+            const [len1, len2] = [...progressbarInputs].map((input) => input.value.length);
+            progressbarInputs[0].setSelectionRange(len1, len1);
```
Comment (brboi) [inline addons/web/static/tests/views/fields/numeric_fields_tests.js:165]: Please explain why is this needed ?

File: addons/web/static/tests/views/fields/numeric_fields_tests.js [hunk 2]
```diff
@@ -180,4 +183,124 @@ QUnit.module("Fields", (hooks) => {
             assert.strictEqual(progressbarInputs[1].value, "0🇧🇪44🇧🇪🇧🇪");
         }
     );
+
+    QUnit.test(
+        "Numeric fields: NumpadDecimal key is different from the decimalPoint",
+        async function (assert) {
+            registry.category("services").remove("localization");
```
Comment (brboi) [inline addons/web/static/tests/views/fields/numeric_fields_tests.js:190]: nitpicking: you can use `force: true` when adding the fake service below.
or maybe it should work to simply patchWithCleanup the localization object from `@web/core/l10n/localization`

no need to change this, just saying to share knowledge ;-)

File: addons/web/static/tests/views/fields/numeric_fields_tests.js [hunk 3]
```diff
@@ -180,4 +183,124 @@ QUnit.module("Fields", (hooks) => {
...
+                el.setSelectionRange(...selectionRange);
+                el.dispatchEvent(new KeyboardEvent("keydown", { code: "NumpadDecimal", key: "." }));
+
+                // dispatch an extra keydown event and assert that it's not default prevented
+                const extraEvent = new KeyboardEvent("keydown", { code: "Digit1", key: "1" });
+                el.dispatchEvent(extraEvent);
+                assert.ok(
+                    !extraEvent.defaultPrevented,
+                    "Keydown event should not be default prevented"
+                );
```
Comment (brboi) [inline addons/web/static/tests/views/fields/numeric_fields_tests.js:244]: if you cant to assert this, you should do it after the below nexttick.
also, by symmetry you should assert the first event is defaultPrevented

File: addons/web/static/tests/views/fields/numeric_fields_tests.js [hunk 4]
```diff
@@ -180,4 +183,124 @@ QUnit.module("Fields", (hooks) => {
...
+                // Selection range should be 1 + the specified selection start.
+                assert.strictEqual(el.selectionStart, selectionRange[0] + 1);
+                assert.strictEqual(el.selectionEnd, selectionRange[0] + 1);
+                await nextTick();
+                assert.strictEqual(el.value, expectedValue, msg || "");
```
Comment (brboi) [inline addons/web/static/tests/views/fields/numeric_fields_tests.js:249]: nitpicking: msg fallback is not needed

File: addons/web/static/tests/views/fields/numeric_fields_tests.js [hunk 5]
```diff
@@ -180,4 +183,124 @@ QUnit.module("Fields", (hooks) => {
...
+                click(document.body);
```
Comment (brboi) [inline addons/web/static/tests/views/fields/numeric_fields_tests.js:250]: what is it for ?

**Review by brboi — CHANGES_REQUESTED** [review]

File: addons/web/static/tests/views/fields/numeric_fields_tests.js (hunk 5 above)
Comment (caburj) [reply addons/web/static/tests/views/fields/numeric_fields_tests.js:250]: It is for removing the focus to the current input. I can remove it because it's not really that important.

File: addons/web/static/tests/views/fields/numeric_fields_tests.js (hunk 2 above)
Comment (caburj) [reply addons/web/static/tests/views/fields/numeric_fields_tests.js:190]: I'll try patchWithCleanup. I kinda like it too.

File: addons/web/static/tests/views/fields/numeric_fields_tests.js (hunk 1 above)
Comment (caburj) [reply addons/web/static/tests/views/fields/numeric_fields_tests.js:165]: When clicking the progressbar, it focuses on the first input but the full value is highlighted, so I need to set the selection to the end of the value so that the "decimalPoint" is appended.

It's the thing we are fixing. Basically the original test is wrong because even if the value is highlighted, the "decimalPoint" is appended.

File: addons/web/static/tests/views/fields/numeric_fields_tests.js (hunk 3 above)
Comment (caburj) [reply addons/web/static/tests/views/fields/numeric_fields_tests.js:244]: Man, even if "preventDefault" is called, the "defaultPrevented" is not set to "true". I patched the preventDefault to perform the proper assertion.

File: addons/web/static/src/views/fields/numpad_decimal_hook.js [hunk 6]
```diff
@@ -11,6 +11,9 @@ const { useRef, useEffect } = owl;
  * reference in the current component. It can be placed directly on an
  * input or an element containing multiple inputs that require the
  * behavior
+ *
+ * NOTE: Special consideration for the input type = "number". In this
+ * case, whatever the user types, we let the browser's default behavior.
```
Comment (caburj) [inline addons/web/static/src/views/fields/numpad_decimal_hook.js:16]: Here's the note about input with type = "number".

File: addons/web/static/tests/views/fields/numeric_fields_tests.js (hunk 1 above)
Comment (brboi) [reply addons/web/static/tests/views/fields/numeric_fields_tests.js:170]: Ha! I see.
While you are at it you should add a comment.
Or maybe don't do that "magic" and assert that the selection range for the progressbar field does select all.

File: addons/web/static/src/views/fields/numpad_decimal_hook.js [hunk 7]
```diff
@@ -35,19 +35,18 @@ export function useNumpadDecimal() {
...
-            return () => {
-                listeners.forEach((input) => input.removeEventListener("keydown", handler));
-            };
-        },
-        () => [ref.el]
-    );
+    useEffect(() => {
+        const el = ref.el;
+        if (el) {
+            const inputs = el.nodeName === "INPUT" ? [el] : el.querySelectorAll("input");
```
Comment (brboi) [inline addons/web/static/src/views/fields/numpad_decimal_hook.js:41]: Maybe you could do this to get rid of the listeners array ?
```suggestion
        let inputs = [];
        if (el) {
            inputs = el.nodeName === "INPUT" ? [el] : el.querySelectorAll("input");
```

If you want, you can also move the `let inputs = [];` in the closure of the useNumpadDecimal hook instead of the useEffect callback, I don't think that it matters a lot.

File: addons/web/static/tests/views/fields/numeric_fields_tests.js [hunk 8]
```diff
@@ -306,4 +310,64 @@ QUnit.module("Fields", (hooks) => {
...
+             */
+            function testInputElements(inputEls) {
+                for (const inputEl of inputEls) {
+                    inputEl.focus();
+                    const numpadDecimalEvent = new KeyboardEvent("keydown", {
+                        code: "NumpadDecimal",
+                        key: ".",
+                    });
+                    numpadDecimalEvent.preventDefault = () => assert.step("preventDefault");
+                    inputEl.dispatchEvent(numpadDecimalEvent);
```
Comment (brboi) [inline addons/web/static/tests/views/fields/numeric_fields_tests.js:332]: though it is not really needed here, I think that - almost - every time we dispatch an event we should await the next tick.

File: addons/web/static/tests/views/fields/numeric_fields_tests.js [hunk 9]
```diff
@@ -306,4 +310,64 @@ QUnit.module("Fields", (hooks) => {
...
+
+                    // dispatch an extra keydown event and assert that it's not default prevented
+                    const extraEvent = new KeyboardEvent("keydown", { code: "Digit1", key: "1" });
+                    extraEvent.preventDefault = () => {
+                        throw new Error("should not be default prevented");
+                    };
+                    inputEl.dispatchEvent(extraEvent);
```
Comment (brboi) [inline addons/web/static/tests/views/fields/numeric_fields_tests.js:339]: same here

File: addons/web/static/tests/views/fields/numeric_fields_tests.js [hunk 10]
```diff
@@ -306,4 +310,64 @@ QUnit.module("Fields", (hooks) => {
             });
         }
     );
+
+    QUnit.test(
+        "useNumpadDecimal should synchronize handlers on input elements",
+        async function (assert) {
+            target = getFixture();
+            patchWithCleanup(localization, { decimalPoint: ",", thousandsSep: "." });
```
Comment (brboi) [inline addons/web/static/tests/views/fields/numeric_fields_tests.js:318]: these two lines are not necessary:
- target is already set in the beforeEach
- the localization service is set in the beforeEach and will already patch the localization data with the same values

File: addons/web/static/tests/views/fields/numeric_fields_tests.js [hunk 11]
```diff
@@ -306,4 +310,64 @@ QUnit.module("Fields", (hooks) => {
...
+                    <input type="text" placeholder="input 1" />
+                    <input t-if="state.showOtherInput" type="text" placeholder="input 2" />
+                </main>
+            `;
+            const comp = await mount(MyComponent, target, { env: await makeTestEnv() });
+
+            // Initially, only one input should be rendered.
+            let inputs = target.querySelectorAll("main > input");
+            assert.strictEqual(inputs.length, 1);
+            testInputElements(inputs);
```
Comment (brboi) [inline addons/web/static/tests/views/fields/numeric_fields_tests.js:361]: you can also do this, but that's up to you ;-)
```suggestion
            assert.containsOnce(target, "main > input");
            testInputElements(target.querySelectorAll("main > input");
```

(same below with `assert.containsN(..., 2)`)

**Review by brboi — CHANGES_REQUESTED** [review]

**Comment by brboi** [comment]
robodoo rebase-ff

**Review by brboi — APPROVED** [review]
robodoo r+
Thanks for this pr 😉
